### ISSUE TITLE
Error while executing SQL command in transaction 'BuiltInTeamInitializer': Must declare the scalar variable "@Json".

### ISSUE THREAD

**Issue by michaelnoonan** [OctopusDeploy/Issues#3734]
In a [recent change](https://github.com/OctopusDeploy/Nevermore/commit/8b66404cb495a0e47790dd89f33bc6760f5a671d#diff-5678119f299dc14913a0e5008a2fbaf3R302) to Nevermore we improved our support of case sensitive SQL Servers, but accidentally introduced this new bug in Octopus `3.16.2`:

```
FATAL  Error while executing SQL command in transaction 'BuiltInTeamInitializer': Must declare the scalar variable "@Json".
The command being executed was:
UPDATE dbo.[UserRole]  SET [Name] = @Name, [JSON] = @Json WHERE Id = @Id
SQL Error 137 - Must declare the scalar variable "@Json".
System.Exception: Error while executing SQL command in transaction 'BuiltInTeamInitializer': Must declare the scalar variable "@Json".
The command being executed was:
UPDATE dbo.[UserRole]  SET [Name] = @Name, [JSON] = @Json WHERE Id = @Id ---> System.Data.SqlClient.SqlException: Must declare the scalar variable "@Json".
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString, Boolean isInternal, Boolean forDescribeParameterEncryption)
   at System.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, Boolean async, Int32 timeout, Task& task, Boolean asyncWrite, Boolean inRetry, SqlDataReader ds, Boolean describeParameterEncryptionRequest)
   at System.Data.SqlClient.SqlCommand.RunExecuteReader(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, String method, TaskCompletionSource`1 completion, Int32 timeout, Task& task, Boolean& usedCache, Boolean asyncWrite, Boolean inRetry)
   at System.Data.SqlClient.SqlCommand.InternalExecuteNonQuery(TaskCompletionSource`1 completion, String methodName, Boolean sendToPipe, Int32 timeout, Boolean& usedCache, Boolean asyncWrite, Boolean inRetry)
   at System.Data.SqlClient.SqlCommand.ExecuteNonQuery()
   at Nevermore.Transient.IDbCommandExtensions.<>c__DisplayClass2_0.<ExecuteNonQueryWithRetry>b__0()
   at Nevermore.Transient.RetryPolicy.ExecuteAction[TResult](Func`1 func)
   at Nevermore.RelationalTransaction.Update[TDocument](TDocument instance, String tableHint, Nullable`1 commandTimeoutSeconds)
   --- End of inner exception stack trace ---
   at Nevermore.RelationalTransaction.Update[TDocument](TDocument instance, String tableHint, Nullable`1 commandTimeoutSeconds)
   at Octopus.Core.Initialization.BuiltIn.BuiltInTeamInitializer.EnsureRoleExists(IRelationalTransaction transaction, BuiltInRole builtInRole)
   at Octopus.Core.Initialization.BuiltIn.BuiltInTeamInitializer.EnsureBuiltInRolesExist(IRelationalTransaction transaction)
   at Octopus.Core.Initialization.BuiltIn.BuiltInTeamInitializer.Initialize(IRelationalStore store)
   at Nevermore.StoreInitializer.Initialize()
   at Octopus.Server.OctopusServerEngine.Start()
   at Octopus.Server.Commands.RunCommand.Start()
   at Octopus.Shared.Startup.AbstractCommand.Start(String[] commandLineArguments, ICommandRuntime commandRuntime, OptionSet commonOptions)
   at Octopus.Shared.Startup.WindowsServiceHost.<>c__DisplayClass1_0.<Run>b__0()
   at Octopus.Shared.Startup.WindowsServiceAdapter.RunService()
```

# Who is affected

Customers who are using a Case Sensitive collation for their SQL Server (set at install time) should wait until this bug is fixed before upgrading to Octopus 3.16.2 or later.

# Workarounds

Unfortunately there are no workarounds apart from [changing the collation of your SQL Server](https://www.mssqltips.com/sqlservertip/3519/changing-sql-server-collation-after-installation/).

# Background

We [don't support case sensitive database collations](https://octopus.com/docs/installation/installing-octopus/sql-server-database-requirements), but it would be bad form to absolutely mandate the SQL Server itself must use a case-insensitive collation.

The reason we care is because command parameters like `@Json` are interpreted using the **SQL Server** collation, not the **Database** collation!

# References

Source: https://secure.helpscout.net/conversation/412620581?folderId=557077
Relates to: https://github.com/OctopusDeploy/Issues/issues/2965
Relates to: https://github.com/OctopusDeploy/Issues/issues/3623

**Comment by michaelnoonan** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [A](https://trello.com/c/tIU9fKGS/648-a)

**Comment by michaelnoonan** [comment]
I'm having trouble closing this in time for the 3.16.3 release since testing is taking a long time!

**Comment by kf-josephmegkousoglou** [comment]
I still seem to have the same issue on 3.16.4

**Comment by michaelnoonan** [comment]
I've got a bigger PR in progress to deal with the entire issue of Server Collation being case-sensitive, but am shipping a more targeted fix as part of `3.16.6`.
